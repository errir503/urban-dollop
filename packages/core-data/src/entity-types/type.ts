/**
 * Internal dependencies
 */
import type { Context, ContextualField, OmitNevers } from './helpers';

import type { BaseEntityRecords as _BaseEntityRecords } from './base-entity-records';

declare module './base-entity-records' {
	export namespace BaseEntityRecords {
		export interface Type< C extends Context > {
			/**
			 * All capabilities used by the post type.
			 */
			capabilities: ContextualField<
				Record< string, string >,
				'edit',
				C
			>;
			/**
			 * A human-readable description of the post type.
			 */
			description: ContextualField< string, 'view' | 'edit', C >;
			/**
			 * Whether or not the post type should have children.
			 */
			hierarchical: ContextualField< boolean, 'view' | 'edit', C >;
			/**
			 * Whether or not the post type can be viewed.
			 */
			viewable: ContextualField< boolean, 'edit', C >;
			/**
			 * Human-readable labels for the post type for various contexts.
			 */
			labels: ContextualField< Record< string, string >, 'edit', C >;
			/**
			 * The title for the post type.
			 */
			name: string;
			/**
			 * An alphanumeric identifier for the post type.
			 */
			slug: string;
			/**
			 * All features, supported by the post type.
			 */
			supports: ContextualField< Record< string, string >, 'edit', C >;
			/**
			 * Taxonomies associated with post type.
			 */
			taxonomies: ContextualField< string[], 'view' | 'edit', C >;
			/**
			 * REST base route for the post type.
			 */
			rest_base: string;
			/**
			 * REST route's namespace for the post type.
			 */
			rest_namespace: string;
			/**
			 * The visibility settings for the post type.
			 */
			visibility: ContextualField< TypeVisibility, 'edit', C >;
		}

		interface TypeVisibility {
			/**
			 * Whether to generate a default UI for managing this post type.
			 */
			show_ui: boolean;
			/**
			 * Whether to make the post type is available for selection in navigation menus.
			 */
			show_in_nav_menus: boolean;
		}
	}
}

export type Type< C extends Context = 'edit' > = OmitNevers<
	_BaseEntityRecords.Type< C >
>;
