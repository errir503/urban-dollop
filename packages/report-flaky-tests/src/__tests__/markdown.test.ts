/**
 * External dependencies
 */
import * as core from '@actions/core';

/**
 * Internal dependencies
 */
import {
	formatTestErrorMessage,
	formatTestResults,
	renderIssueBody,
	parseFormattedTestResults,
	parseIssueBody,
	renderCommitComment,
	isReportComment,
} from '../markdown';
import type { ReportedIssue } from '../types';

jest.useFakeTimers( 'modern' ).setSystemTime( new Date( '2020-05-10' ) );

jest.spyOn( core, 'error' ).mockImplementation( () => {} );

describe( 'formatTestErrorMessage', () => {
	it( 'should format test error message for jest-circus', async () => {
		const { default: flakyTest } = await import(
			'../__fixtures__/Should insert new template part on creation.json'
		);

		const formatted = formatTestErrorMessage( flakyTest as any );

		expect( formatted ).toMatchSnapshot();
	} );

	it( 'should format test error message for @playwright/test', async () => {
		const { default: flakyTest } = await import(
			'../__fixtures__/should copy only partial selection of text blocks.json'
		);

		const formatted = formatTestErrorMessage( flakyTest as any );

		expect( formatted ).toMatchSnapshot();
	} );
} );

describe( 'formatTestResults', () => {
	it( 'should format flaky test results without error messages', () => {
		const formatted = formatTestResults( {
			date: new Date(),
			failedTimes: 1,
			headBranch: 'trunk',
			runURL: 'https://github.com/WordPress/gutenberg/actions/runs/2282393879',
		} );

		expect( formatted ).toMatchInlineSnapshot(
			`"<!-- __TEST_RESULT__ --><time datetime="2020-05-10T00:00:00.000Z"><code>[2020-05-10T00:00:00.000Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/WordPress/gutenberg/actions/runs/2282393879"><code>trunk</code></a>.<!-- /__TEST_RESULT__ -->"`
		);

		expect( renderToDisplayText( formatted ) ).toMatchInlineSnapshot(
			`"[2020-05-10T00:00:00.000Z] Test passed after 1 failed attempt on trunk."`
		);
	} );

	it( 'should format flaky test results with error messages', () => {
		const formatted = formatTestResults( {
			date: new Date(),
			failedTimes: 1,
			headBranch: 'trunk',
			runURL: 'https://github.com/WordPress/gutenberg/actions/runs/2282393879',
			errorMessage: 'Error:\n  Oops...\n',
		} );

		expect( formatted ).toMatchInlineSnapshot( `
		"<!-- __TEST_RESULT__ --><details>
		<summary>
			<time datetime="2020-05-10T00:00:00.000Z"><code>[2020-05-10T00:00:00.000Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/WordPress/gutenberg/actions/runs/2282393879"><code>trunk</code></a>.
		</summary>

		\`\`\`
		Error:
		  Oops...

		\`\`\`
		</details><!-- /__TEST_RESULT__ -->"
	` );
	} );
} );

describe( 'renderIssueBody', () => {
	it( 'should render issue body', () => {
		const meta = {
			failedTimes: 5,
			totalCommits: 100,
		};

		const body = renderIssueBody( {
			meta,
			testTitle: 'Test title',
			testPath: 'test/e2e/specs/test-title.spec.js',
			formattedTestResults: 'FORMATTED_TEST_RESULTS',
		} );

		expect( body ).toMatchInlineSnapshot( `
		"<!-- __META_DATA__:{"failedTimes":5,"totalCommits":100} -->
		**Flaky test detected. This is an auto-generated issue by GitHub Actions. Please do NOT edit this manually.**

		## Test title
		Test title

		## Test path
		\`test/e2e/specs/test-title.spec.js\`

		## Errors
		<!-- __TEST_RESULTS_LIST__ -->
		FORMATTED_TEST_RESULTS
		<!-- /__TEST_RESULTS_LIST__ -->
		"
	` );
	} );
} );

describe( 'parseFormattedTestResults', () => {
	it( 'should parse formatted test results without error messages', () => {
		const testResult = {
			date: new Date(),
			failedTimes: 1,
			headBranch: 'trunk',
			runURL: 'https://github.com/WordPress/gutenberg/actions/runs/2297863316',
		};

		const formatted = formatTestResults( testResult );

		const parsed = parseFormattedTestResults( formatted );

		expect( parsed ).toEqual( testResult );
	} );

	it( 'should parse formatted test results with error messages', () => {
		const testResult = {
			date: new Date(),
			failedTimes: 1,
			headBranch: 'trunk',
			runURL: 'https://github.com/WordPress/gutenberg/actions/runs/2297863316',
			errorMessage: 'Error:\n  Oops...\n',
		};

		const formatted = formatTestResults( testResult );

		const parsed = parseFormattedTestResults( formatted );

		expect( parsed ).toEqual( testResult );
	} );

	it( 'should handle GitHub converted line breaks (\r\n) in the result', () => {
		const testResult = {
			date: new Date(),
			failedTimes: 1,
			headBranch: 'trunk',
			runURL: 'https://github.com/WordPress/gutenberg/actions/runs/2297863316',
			errorMessage: 'Error:\n  Oops...\n',
		};

		const formatted = formatTestResults( testResult ).replace(
			/\n/g,
			'\r\n'
		);

		const parsed = parseFormattedTestResults( formatted );

		expect( parsed ).toEqual( testResult );
	} );
} );

describe( 'parseIssueBody', () => {
	it( 'should parse issue body', async () => {
		const testResults = [
			formatTestResults( {
				date: new Date( '2020-08-01' ),
				failedTimes: 1,
				headBranch: 'trunk',
				runURL: 'https://github.com/WordPress/gutenberg/actions/runs/2282393879',
			} ),
			formatTestResults( {
				date: new Date(),
				failedTimes: 2,
				headBranch: 'try/some-branch',
				runURL: 'https://github.com/WordPress/gutenberg/actions/runs/2297863316',
				errorMessage: 'Error:\n  Oops...\n  Oops again',
			} ),
			`\n<!-- __TEST_RESULT__ -->Some broken log.\nOops.\n<!-- /__TEST_RESULT__ -->\n`,
		];

		const meta = {
			failedTimes: 5,
			totalCommits: 95,
		};

		const body = renderIssueBody( {
			meta,
			testTitle: 'Test title',
			testPath: 'test/e2e/specs/test-title.spec.js',
			formattedTestResults: testResults.join( '\n' ),
		} );

		expect( body ).toMatchSnapshot();

		const parsed = parseIssueBody( body );

		expect( core.error ).toHaveBeenCalledTimes( 1 );

		expect( parsed ).toEqual( {
			meta,
			testResults: [
				{
					date: new Date( '2020-08-01' ),
					failedTimes: 1,
					headBranch: 'trunk',
					runURL: 'https://github.com/WordPress/gutenberg/actions/runs/2282393879',
				},
				{
					date: new Date(),
					failedTimes: 2,
					headBranch: 'try/some-branch',
					runURL: 'https://github.com/WordPress/gutenberg/actions/runs/2297863316',
					errorMessage: 'Error:\n  Oops...\n  Oops again',
				},
			],
		} );
	} );
} );

describe( 'renderCommitComment', () => {
	it( 'render the commit comment', () => {
		const runURL = 'runURL';
		const reportedIssues: ReportedIssue[] = [
			{
				testTitle: 'title1',
				testPath: 'path1',
				issueNumber: 1,
				issueUrl: 'url1',
			},
			{
				testTitle: 'title2',
				testPath: 'path2',
				issueNumber: 2,
				issueUrl: 'url2',
			},
		];
		const commitSHA = 'commitSHA';

		const commentBody = renderCommitComment( {
			reportedIssues,
			runURL,
			commitSHA,
		} );

		expect( commentBody ).toMatchInlineSnapshot( `
		"<!-- flaky-tests-report-comment -->
		**Flaky tests detected in commitSHA.**
		Some tests passed with failed attempts. The failures may not be related to this commit but are still reported for visibility. See [the documentation](https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/testing-overview.md#flaky-tests) for more information.

		🔍  Workflow run URL: runURL
		📝  Reported issues:
		- #1 in \`path1\`
		- #2 in \`path2\`"
	` );
	} );
} );

describe( 'isReportComment', () => {
	it( 'matches the report comment', () => {
		const commentBody = renderCommitComment( {
			reportedIssues: [],
			runURL: '',
			commitSHA: 'commitSHA',
		} );

		expect( isReportComment( commentBody ) ).toBe( true );

		expect( isReportComment( 'random string' ) ).toBe( false );
	} );
} );

function renderToDisplayText( html: string ) {
	const container = document.createElement( 'div' );
	container.innerHTML = html;
	return container.textContent!.replace( /\s+/g, ' ' );
}
