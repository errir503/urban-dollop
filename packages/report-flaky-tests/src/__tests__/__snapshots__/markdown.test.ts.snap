// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatTestErrorMessage should format test error message for @playwright/test 1`] = `
"Error: Snapshot comparison failed:

  <!-- wp:paragraph -->
  <p>block</p>
  <!-- /wp:paragraph -->

  <!-- wp:paragraph -->
  <p>B </p>
  <!-- /wp:paragraph -->

  <!-- wp:paragraph -->
  <p>A block</p>
  <!-- /wp:paragraph -->

  <!-- wp:paragraph -->
  <p></p>
  <!-- /wp:paragraph -->

  <!-- wp:paragraph -->
  <p>B block</p>
  <!-- /wp:paragraph -->

Expected: /home/runner/work/gutenberg/gutenberg/artifacts/test-results/editor-various-copy-cut-paste-Copy-cut-paste-should-copy-only-partial-selection-of-text-blocks-chromium/Copy-cut-paste-should-copy-only-partial-selection-of-text-blocks-2-expected.txt
Received: /home/runner/work/gutenberg/gutenberg/artifacts/test-results/editor-various-copy-cut-paste-Copy-cut-paste-should-copy-only-partial-selection-of-text-blocks-chromium/Copy-cut-paste-should-copy-only-partial-selection-of-text-blocks-2-actual.txt
    at /home/runner/work/gutenberg/gutenberg/test/e2e/specs/editor/various/copy-cut-paste.spec.js:253:52"
`;

exports[`formatTestErrorMessage should format test error message for jest-circus 1`] = `
"  ● Template Part › Template part block › Template part placeholder › Should insert new template part on creation

    expect(jest.fn()).not.toHaveErrored(expected)

    Expected mock function not to be called but it was called with:
    ["TypeError: Cannot read properties of null (reading 'frameElement')

      at Vo (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=dfd3a79ce1dc54c31b6ed591bcb0d55a:3:21925)
      at ft (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=6.1-alpha-53362:1:43451)
      at Wt (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=6.1-alpha-53362:1:50270)
      at ts (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=6.1-alpha-53362:1:112276)
      at Fr (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=6.1-alpha-53362:1:77758)
      at Dr (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=6.1-alpha-53362:1:77686)
      at Rr (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=6.1-alpha-53362:1:77549)
      at Nr (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=6.1-alpha-53362:1:74544)
      at ../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=6.1-alpha-53362:1:30170
      at unstable_runWithPriority (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react.min.js?ver=6.1-alpha-53362:1:7430)"]
          at runMicrotasks (<anonymous>)

  ● Template Part › Template part block › Template part placeholder › Should insert new template part on creation

    expect(jest.fn()).not.toHaveErrored(expected)

    Expected mock function not to be called but it was called with:
    ["TypeError: Cannot read properties of null (reading 'frameElement')

      at Vo (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=dfd3a79ce1dc54c31b6ed591bcb0d55a:3:21925)
      at ft (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=6.1-alpha-53362:1:43451)
      at Wt (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=6.1-alpha-53362:1:50270)
      at ts (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=6.1-alpha-53362:1:112276)
      at Fr (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=6.1-alpha-53362:1:77758)
      at Dr (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=6.1-alpha-53362:1:77686)
      at Rr (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=6.1-alpha-53362:1:77549)
      at Nr (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=6.1-alpha-53362:1:74544)
      at ../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=6.1-alpha-53362:1:30170
      at unstable_runWithPriority (../../http:/localhost:8889/wp-content/plugins/gutenberg/build/vendors/react.min.js?ver=6.1-alpha-53362:1:7430)"]
          at runMicrotasks (<anonymous>)
"
`;

exports[`parseIssueBody should parse issue body 1`] = `
"<!-- __META_DATA__:{"failedTimes":5,"totalCommits":95} -->
**Flaky test detected. This is an auto-generated issue by GitHub Actions. Please do NOT edit this manually.**

## Test title
Test title

## Test path
\`test/e2e/specs/test-title.spec.js\`

## Errors
<!-- __TEST_RESULTS_LIST__ -->
<!-- __TEST_RESULT__ --><time datetime="2020-08-01T00:00:00.000Z"><code>[2020-08-01T00:00:00.000Z]</code></time> Test passed after 1 failed attempt on <a href="https://github.com/WordPress/gutenberg/actions/runs/2282393879"><code>trunk</code></a>.<!-- /__TEST_RESULT__ -->
<!-- __TEST_RESULT__ --><details>
<summary>
	<time datetime="2020-05-10T00:00:00.000Z"><code>[2020-05-10T00:00:00.000Z]</code></time> Test passed after 2 failed attempts on <a href="https://github.com/WordPress/gutenberg/actions/runs/2297863316"><code>try/some-branch</code></a>.
</summary>

\`\`\`
Error:
  Oops...
  Oops again
\`\`\`
</details><!-- /__TEST_RESULT__ -->

<!-- __TEST_RESULT__ -->Some broken log.
Oops.
<!-- /__TEST_RESULT__ -->

<!-- /__TEST_RESULTS_LIST__ -->
"
`;
